### PR TITLE
Random encounter 1

### LINKED ISSUE #56601: I still can't remove NPCs from the map
### Is your feature request related to a problem? Please describe.

I'm working on making random encounters, but there's still no way to despawn an NPC. I can send the NPC to walk off on their way, or teleport them to another location, but in the context of a random map, that doesn't really work for two reasons:
- I don't want the NPC to be persistent and show up again later, necessarily: I can track persistence with global variables if I want an NPC to recur.
- the location of the random encounter is random, I can't guarantee a target terrain is going to be in range, and I am sure we'll get errors if I try to teleport or walk the random NPC to an overmap it can't locate.
Overall as well, we just should have the ability to add and remove things from the map, for basic scripting reasons.

### Solution you would like.

overmapbuffer.cpp contains a function, `shared_ptr_fast<npc> overmapbuffer::remove_npc( const character_id &id )`
I believe this can be used to remove an NPC. All I want is the ability to define an `effect` that calls the character_id and removes it, without killing it or causing a map message. This should be very easy to do, but I don't know how.

Syntax like:
```
"effects": {
  "remove_npc": "guard"
}
```
This would then just call that function and do it as that function does it. Viola.

### Describe alternatives you have considered.

I could have the NPC walk to another location, but this would probably result in them dying randomly somewhere.

I could try to teleport the NPC to a relevant location like its parent faction base, but I am almost certain this will produce errors if that location isn't loaded yet.

I could, and I'm not joking this is currently my best option, kill the NPC and delete the corpse and all its items. This will require a lot of json script for what I believe is a very small amount of code.

### Additional context

Please, please help me.

### PR DESCRIPTION
**Pull request by I-am-Erk**
#### Summary
Content "Adds a prototypic random encounter with free merchants at a public washroom"

#### Purpose of change
The world feels more dead than it should: although we can run into vague wandering NPCs, there is a demand for more organized stabilized NPC travel. The classic example is that free merchant caravans, bandit gangs, and various sorts of scouts and scavengers from the faction should show up at various times.

This adds a prototype example, a handful of merchants and guards in a pickup truck.

#### Describe the solution
Adds two merchant guards and a trader that spawn with a pickup truck at roadside public washrooms, starting a week after the game starts. They won't spawn at night or in winter, and you can only run into them once per day. After an hour or so, they despawn once you're outside the reality bubble.

Adds a script by @Mylie to remove NPCs from the map. This script only works when the player is outside the reality bubble. Closes #56601.

Things I would like to add in this PR or a followup:
- inventory adjustments. The trader has almost nothing worth trading right now and no weapons.
- custom classes for the guards, again making them more durable to reduce the chance of the caravan getting wiped out by random zombies.

#### Describe alternatives you've considered
I considered adding persistent NPCs that remember you, but that seems like a next step.
I considered adding a free merchant scout instead, and that should still be a thing.

#### Testing
merchants spawn as advertised and despawn as advertised. Dialogue and directions run without errors

There is a likely bug: if the despawn timer hits while you're nearby, the pickup truck may despawn before the NPCs do. We will probably need to refine the conditionals around despawning for this to make sure they run synced. That's beyond my skill and could be considered a block to merging this. Or, someone who knows stuff could expand the `remove_npc` script to also work inside the bubble.

#### Additional context
This framework could trivially be used to add some great stuff:
- We should add similar caravaners and mercenaries to the back room of the refugee center, showing up sometimes and leaving. Likewise the same framework can be used to have the Arsonist and Merc persistent characters, and Yusuke the beggar, be around sometimes but gone scavenging/running jobs others.
- We should add scavengers that are sometimes found trading in the front room of the center.
- Other factions should have moving random encounters and caravans should have a chance of showing up in other places. Once we do that, the chance of running into this specific encounter at the rest stop should drop.
- With a bit more infrastructure we could add a "trade depot" base structure for player bases and dialogue with the free merchant teamster to arrange caravans to come to your base sometimes.

In the future we might add one or two "upgraded" versions of the caravan, with a reinforced truck and maybe a gun turret, better gear on the guards, and some later-cataclysm dialogue.

### CODE REVIEW

File: data/json/encounters/randenc_caravans.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        }
+      ]
+    }
+  },
+  {
+    "type": "mapgen",
+    "method": "json",
+    "update_mapgen_id": "nest_RandEnc_roadstop_a_remove",
+    "//": "removes a random encounter at this roadstop",
+    "object": { "remove_vehicles": [ { "vehicles": [ "pickup" ], "x": [ 1, 15 ], "y": [ 1, 10 ] } ] }
```
Comment (Night-Pryanik) [inline data/json/encounters/randenc_caravans.json:67]: Could you please explain why `place_vehicles` has strict `x` and `y` coordinates, but `remove_vehicles` has an array of coordinates?

File: data/json/npcs/random_encounters/free_merchant_caravans.json [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    "responses": [
+      { "text": "Do you have anything to trade?", "topic": "TALK_CARAVAN_MERCHANT_Trade", "effect": "start_trade" },
+      { "text": "What's your story?", "topic": "TALK_CARAVAN_MERCHANT_Story" },
+      { "text": "Nothing much.  Just passing through.", "topic": "TALK_DONE" }
+    ]
+  },
+  {
+    "type": "talk_topic",
+    "id": "TALK_CARAVAN_MERCHANT_Threatening",
+    "dynamic_line": "No, not at all. Just being cautious.  You know how it gets out here.  We're not looking for trouble.  In fact, if you've got anything worthwhile, maybe we could do a little trading.",
```
Comment (Night-Pryanik) [inline data/json/npcs/random_encounters/free_merchant_caravans.json:59]:
```suggestion
    "dynamic_line": "No, not at all.  Just being cautious.  You know how it gets out here.  We're not looking for trouble.  In fact, if you've got anything worthwhile, maybe we could do a little trading.",
```

File: data/json/npcs/random_encounters/free_merchant_caravans.json [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+          ]
+        }
+      },
+      { "text": "Thanks for chatting.  I'd better go.", "topic": "TALK_DONE" }
+    ]
+  },
+  {
+    "type": "talk_topic",
+    "id": "TALK_CARAVAN_MERCHANT_Home",
+    "dynamic_line": "Here, I'll mark it on your map. Feel free to come by and trade, or if you're the type, we might have use for your services as a merc.",
```
Comment (Night-Pryanik) [inline data/json/npcs/random_encounters/free_merchant_caravans.json:125]:
```suggestion
    "dynamic_line": "Here, I'll mark it on your map.  Feel free to come by and trade, or if you're the type, we might have use for your services as a merc.",
```

File: data/json/npcs/random_encounters/free_merchant_caravans.json [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    "type": "talk_topic",
+    "id": "TALK_CARAVAN_MERCHANT_Trade",
+    "dynamic_line": "Always good to do a little trading.",
+    "responses": [
+      {
+        "text": "Actually I had a little more to trade still.",
+        "topic": "TALK_CARAVAN_MERCHANT_Trade",
+        "effect": "start_trade"
+      },
+      { "text": "Thanks. What were we talking about before?", "topic": "TALK_NONE" },
```
Comment (Night-Pryanik) [inline data/json/npcs/random_encounters/free_merchant_caravans.json:146]:
```suggestion
      { "text": "Thanks.  What were we talking about before?", "topic": "TALK_NONE" },
```

File: data/json/npcs/random_encounters/free_merchant_caravans.json [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    "responses": [
+      { "text": "What are you doing here?", "topic": "TALK_CARAVAN_GUARD_A_Job" },
+      { "text": "You looking for work?  I might be hiring.", "topic": "TALK_CARAVAN_GUARD_A_Hire" },
+      { "text": "Nope, just saying hi.", "topic": "TALK_DONE" }
+    ]
+  },
+  {
+    "type": "talk_topic",
+    "id": "TALK_CARAVAN_GUARD_A_Job",
+    "dynamic_line": "I'm a caravan guard, working for the free merchants.  I'm an independent contractor though.  Seems like a good way to get safety in numbers and try to find a safe place to settle down.",
```
Comment (Night-Pryanik) [inline data/json/npcs/random_encounters/free_merchant_caravans.json:173]: Capitalize Free Merchants? Or it's intended?

File: data/json/encounters/randenc_caravans.json (hunk 1 above)
Comment (I-am-Erk) [reply data/json/encounters/randenc_caravans.json:67]: No strong reason, but it allows different shaped encounters to go in this area. I don't plan on having them all look identical each time.

**Comment by andrei8l** [comment]
I don't think spawning ephemeral NPCs is a good mechanic, unless it's justified by lore somehow (ex: dimensional travelers). If there's no valid place to send them, can't they be sent to the nearest uninitialized overmap?

File: data/json/encounters/randenc_caravans.json (hunk 1 above)
Comment (Night-Pryanik) [reply data/json/encounters/randenc_caravans.json:67]: I mean, according to the code, pickup will be placed at one fixed point every encounter, right?

**Comment by I-am-Erk** [comment]
> I don't think spawning ephemeral NPCs is a good mechanic, unless it's justified by lore somehow (ex: dimensional travelers). If there's no valid place to send them, can't they be sent to the nearest uninitialized overmap?

Once NPCs have the ability to drive, it would be nice to replace 'despawn' with 'get in the truck and drive a few OMT away from the player' before despawning', but since that's a feature we want anyway, I don't see this blocking it.

Eventually I would like to see more persistent trader NPCs that remember you, but adding the requirements for that will take longer, and we'll still need the random ones as a potential fallback and to fill out the ranks.

File: data/json/encounters/randenc_caravans.json (hunk 1 above)
Comment (I-am-Erk) [reply data/json/encounters/randenc_caravans.json:68]: For the moment yes, but not forever. Also it's possible for the truck to move during gameplay but still be in a situation where we want it to despawn, eg. if the player nudges it with another vehicle.

File: data/json/encounters/randenc_caravans.json [hunk 6]
```diff
@@ -0,0 +1,70 @@
+[
+  {
+    "type": "effect_on_condition",
+    "id": "EOC_RandEnc_Roadstop_add",
```
Comment (Maleclypse) [inline data/json/encounters/randenc_caravans.json:4]: Do we need any kind of tests setup to make sure we don't introduce regressions to these specific sets of eoc's?   Or are eoc's isolated enough that shouldn't be an issue?

**Comment by kevingranade** [comment]
> I don't think spawning ephemeral NPCs is a good mechanic

This is definitely not the desired long-term way for this to work, but resolving this "properly" is *really* far over the horizon right now, so I'm loath to block content that is at least representative of how it's supposed to work.

**Comment by I-am-Erk** [comment]
The key thing I hope to add to the encounter framework pretty soon is a kind of "population" variable that downticks if you kill faction members. So, the more random free merchants you murder, the fewer random free merchants you'll encounter for example. I'll probably put that in fairly soon. That will keep us from needing to model movements of merchants, but still allow them to be wiped out.

**Comment by LucasBurrns** [comment]
Is it planned to track those encounters on the overmap? Roughly saying to be able to intercept the travelling entity if the path is known, or encounter of entities on the overmap. I'm not talking about any precise pathfinding though, at most - travel by road if possible.

**Comment by I-am-Erk** [comment]
Probably not any time soon.

**Pull request merged**